File: scripts/test-bank-matrix.py
#!/usr/bin/env python3
"""
Top-bank FinTS anonymous-init smoke matrix.

What it tests (no credentials required):
- Endpoint reachability + FinTS envelope acceptance
- BPD presence (HIBPA)
- HIKAZS / HICAZS support advertisement
- HITANS TAN method IDs

Usage:
    python3 scripts/test-bank-matrix.py
    python3 scripts/test-bank-matrix.py --timeout 20
"""

from __future__ import annotations

import argparse
import base64
import dataclasses
import json
import re
from pathlib import Path
from typing import Dict, List, Optional, Tuple

import requests

PRODUCT_ID = "F7C4049477F6136957A46EC28"
COUNTRY = "280"


@dataclasses.dataclass
class BankInfo:
    blz: str
    bic: str
    name: str
    url: str


@dataclasses.dataclass
class ProbeResult:
    label: str
    blz: str
    name: str
    url: str
    endpoint_ok: bool
    bpd_ok: bool
    structural_ok: bool
    hikaz_versions: List[str]
    has_hicazs: bool
    tan_methods: List[str]
    response_codes: List[str]
    http_status: Optional[int]
    error: str
    variant: str


def build_anonymous_init(blz: str) -> bytes:
    hkidn = f"HKIDN:2:2+{COUNTRY}:{blz}+0+0+0'"
    hkvvb = f"HKVVB:3:3+0+0+0+{PRODUCT_ID}+1.0'"
    inner = hkidn + hkvvb
    hnhbs = "HNHBS:4:1+1'"

    header_prefix = "HNHBK:1:3+"
    header_suffix = "+300+0+1'"
    header_len = len(header_prefix) + 12 + len(header_suffix)
    total = header_len + len(inner) + len(hnhbs)
    header = f"{header_prefix}{total:012d}{header_suffix}"
    msg = header + inner + hnhbs
    return msg.encode("iso-8859-1")


def build_anonymous_init_with_hksyn(blz: str) -> bytes:
    hkidn = f"HKIDN:2:2+{COUNTRY}:{blz}+0+0+0'"
    hkvvb = f"HKVVB:3:3+0+0+0+{PRODUCT_ID}+5.0.0'"
    hksyn = "HKSYN:4:3+0'"
    inner = hkidn + hkvvb + hksyn
    hnhbs = "HNHBS:5:1+1'"

    header_prefix = "HNHBK:1:3+"
    header_suffix = "+300+0+1'"
    header_len = len(header_prefix) + 12 + len(header_suffix)
    total = header_len + len(inner) + len(hnhbs)
    header = f"{header_prefix}{total:012d}{header_suffix}"
    msg = header + inner + hnhbs
    return msg.encode("iso-8859-1")


def decode_fints_response(raw_bytes: bytes) -> str:
    raw = raw_bytes.decode("iso-8859-1", errors="replace")
    clean = raw.replace("\n", "").replace("\r", "").replace(" ", "")
    decoded = base64.b64decode(clean)
    return decoded.decode("iso-8859-1", errors="replace")


def parse_segment_version(segment_header: str) -> Optional[str]:
    # Segment header format e.g. HIKAZS:4:7
    parts = segment_header.split(":")
    if len(parts) < 3:
        return None
    version = parts[2]
    return version if version.isdigit() else None


def analyze_response(decoded_text: str) -> Tuple[bool, bool, List[str], bool, List[str], List[str]]:
    segments = [s for s in decoded_text.split("'") if s]

    bpd_ok = any(seg.startswith("HIBPA") for seg in segments)
    code_list = re.findall(r"([0-9]{4}):", decoded_text)
    codes = sorted(set(code_list))
    structural_ok = "9110" not in codes

    hikaz_versions: List[str] = []
    has_hicazs = False
    tan_methods: set[str] = set()

    for seg in segments:
        if seg.startswith("HIKAZS:"):
            header = seg.split("+", 1)[0]
            ver = parse_segment_version(header)
            if ver is not None:
                hikaz_versions.append(ver)
        elif seg.startswith("HICAZS:"):
            has_hicazs = True
        elif seg.startswith("HITANS:"):
            body = seg.split("+")[1:]  # skip HITANS header
            for field in body:
                m = re.match(r"^([0-9]{3}):[0-9A-Z]:", field)
                if m:
                    tan_methods.add(m.group(1))

    hikaz_versions = sorted(set(hikaz_versions), key=int)
    tan_methods_sorted = sorted(tan_methods)
    return bpd_ok, structural_ok, hikaz_versions, has_hicazs, tan_methods_sorted, codes


def send_anonymous_probe(bank: BankInfo, timeout_s: float, variant: str = "anon") -> ProbeResult:
    try:
        if variant == "anon_hksyn":
            msg = build_anonymous_init_with_hksyn(bank.blz)
        else:
            msg = build_anonymous_init(bank.blz)
        body = base64.b64encode(msg)
        resp = requests.post(
            bank.url,
            data=body,
            headers={"Content-Type": "text/plain"},
            timeout=timeout_s,
        )

        if resp.status_code != 200:
            return ProbeResult(
                label=bank.name,
                blz=bank.blz,
                name=bank.name,
                url=bank.url,
                endpoint_ok=False,
                bpd_ok=False,
                structural_ok=False,
                hikaz_versions=[],
                has_hicazs=False,
                tan_methods=[],
                response_codes=[],
                http_status=resp.status_code,
                error=f"HTTP {resp.status_code}",
                variant=variant,
            )

        decoded = decode_fints_response(resp.content)
        bpd_ok, structural_ok, hikaz_versions, has_hicazs, tan_methods, response_codes = analyze_response(decoded)
        return ProbeResult(
            label=bank.name,
            blz=bank.blz,
            name=bank.name,
            url=bank.url,
            endpoint_ok=True,
            bpd_ok=bpd_ok,
            structural_ok=structural_ok,
            hikaz_versions=hikaz_versions,
            has_hicazs=has_hicazs,
            tan_methods=tan_methods,
            response_codes=response_codes,
            http_status=resp.status_code,
            error="",
            variant=variant,
        )
    except Exception as exc:  # noqa: BLE001
        return ProbeResult(
            label=bank.name,
            blz=bank.blz,
            name=bank.name,
            url=bank.url,
            endpoint_ok=False,
            bpd_ok=False,
            structural_ok=False,
            hikaz_versions=[],
            has_hicazs=False,
            tan_methods=[],
            response_codes=[],
            http_status=None,
            error=str(exc),
            variant=variant,
        )


def choose_better_result(primary: ProbeResult, secondary: ProbeResult) -> ProbeResult:
    # Prefer any successful endpoint response over transport error.
    if not primary.endpoint_ok and secondary.endpoint_ok:
        return secondary
    if primary.endpoint_ok and not secondary.endpoint_ok:
        return primary

    # Prefer richer protocol result: BPD + structural success + more discovered fields.
    primary_score = (
        int(primary.bpd_ok) * 4
        + int(primary.structural_ok) * 2
        + int(bool(primary.hikaz_versions))
        + int(primary.has_hicazs)
    )
    secondary_score = (
        int(secondary.bpd_ok) * 4
        + int(secondary.structural_ok) * 2
        + int(bool(secondary.hikaz_versions))
        + int(secondary.has_hicazs)
    )
    return secondary if secondary_score > primary_score else primary


def parse_banks_zig(path: Path) -> Dict[str, BankInfo]:
    text = path.read_text(encoding="utf-8")
    # makeBank("BLZ", "BIC", "Name", "URL")
    pattern = re.compile(
        r'makeBank\("(?P<blz>\d{8})",\s*"(?P<bic>[^"]*)",\s*"(?P<name>[^"]*)",\s*"(?P<url>[^"]*)"\)'
    )
    out: Dict[str, BankInfo] = {}
    for m in pattern.finditer(text):
        blz = m.group("blz")
        out[blz] = BankInfo(
            blz=blz,
            bic=m.group("bic"),
            name=m.group("name"),
            url=m.group("url"),
        )
    return out


def choose_atruvia_representative(banks_by_blz: Dict[str, BankInfo]) -> Optional[BankInfo]:
    candidates = [
        b
        for b in banks_by_blz.values()
        if "fints1.atruvia.de/cgi-bin/hbciservlet" in b.url
        and ("Volksbank" in b.name or "Raiffeisenbank" in b.name or "VR Bank" in b.name)
    ]
    if not candidates:
        return None
    candidates.sort(key=lambda b: b.blz)
    return candidates[0]


def build_target_list(banks_by_blz: Dict[str, BankInfo]) -> List[Tuple[str, BankInfo]]:
    targets: List[Tuple[str, str]] = [
        ("Comdirect", "20041177"),
        ("Berliner Sparkasse", "10050000"),
        ("Deutsche Bank", "10070000"),
        ("Commerzbank", "10040000"),
        ("Postbank", "10010010"),
        ("ING", "50010517"),
        ("DKB", "12030000"),
    ]

    out: List[Tuple[str, BankInfo]] = []
    for label, blz in targets:
        bank = banks_by_blz.get(blz)
        if bank is not None:
            out.append((label, bank))

    atruvia = choose_atruvia_representative(banks_by_blz)
    if atruvia is not None:
        out.append(("Atruvia representative", atruvia))
    return out


def yes_no(value: bool) -> str:
    return "YES" if value else "NO"


def print_matrix(results: List[ProbeResult]) -> None:
    cols = [
        ("Bank", 24),
        ("BLZ", 8),
        ("Endpoint", 8),
        ("BPD", 4),
        ("Struct", 6),
        ("HIKAZS", 8),
        ("HICAZS", 6),
        ("TAN methods", 18),
        ("Codes", 20),
        ("Variant", 10),
    ]
    header = "  ".join(name.ljust(width) for name, width in cols)
    sep = "  ".join("-" * width for _, width in cols)
    print(header)
    print(sep)
    for r in results:
        hikaz = ",".join(r.hikaz_versions) if r.hikaz_versions else "-"
        tan = ",".join(r.tan_methods) if r.tan_methods else "-"
        codes = ",".join(r.response_codes) if r.response_codes else "-"
        row = [
            r.label[:24].ljust(24),
            r.blz.ljust(8),
            yes_no(r.endpoint_ok).ljust(8),
            yes_no(r.bpd_ok).ljust(4),
            yes_no(r.structural_ok).ljust(6),
            hikaz[:8].ljust(8),
            yes_no(r.has_hicazs).ljust(6),
            tan[:18].ljust(18),
            codes[:20].ljust(20),
            r.variant[:10].ljust(10),
        ]
        print("  ".join(row))
        if r.error:
            print(f"    error: {r.error}")
        print(f"    url: {r.url}")


def main() -> None:
    parser = argparse.ArgumentParser(description="FinTS top-bank anonymous-init matrix")
    parser.add_argument("--timeout", type=float, default=15.0, help="HTTP timeout in seconds per bank")
    parser.add_argument(
        "--try-hksyn-variant",
        action="store_true",
        help="if anon probe is weak, also try anon+HKSYN variant and keep the better result",
    )
    parser.add_argument(
        "--json-out",
        default="scripts/bank-matrix-last.json",
        help="path to write JSON report",
    )
    args = parser.parse_args()

    repo_root = Path(__file__).resolve().parents[1]
    banks_path = repo_root / "libwimg" / "src" / "banks.zig"
    banks_by_blz = parse_banks_zig(banks_path)
    targets = build_target_list(banks_by_blz)

    print(f"Running anonymous FinTS probes for {len(targets)} banks (timeout={args.timeout:.1f}s)")
    print()

    results: List[ProbeResult] = []
    for label, bank in targets:
        print(f"- probing {label} ({bank.blz}) ...")
        primary = send_anonymous_probe(
            BankInfo(blz=bank.blz, bic=bank.bic, name=label, url=bank.url),
            timeout_s=args.timeout,
            variant="anon",
        )
        chosen = primary
        if args.try_hksyn_variant and (not primary.bpd_ok or not primary.structural_ok):
            print("  -> trying anon_hksyn variant ...")
            secondary = send_anonymous_probe(
                BankInfo(blz=bank.blz, bic=bank.bic, name=label, url=bank.url),
                timeout_s=args.timeout,
                variant="anon_hksyn",
            )
            chosen = choose_better_result(primary, secondary)
        results.append(chosen)

    print()
    print_matrix(results)

    endpoint_ok = sum(1 for r in results if r.endpoint_ok)
    bpd_ok = sum(1 for r in results if r.bpd_ok)
    structural_ok = sum(1 for r in results if r.structural_ok)
    print()
    print(
        f"Summary: endpoint_ok={endpoint_ok}/{len(results)}, "
        f"bpd_ok={bpd_ok}/{len(results)}, structural_ok={structural_ok}/{len(results)}"
    )

    json_path = Path(args.json_out)
    payload = {
        "summary": {
            "count": len(results),
            "endpoint_ok": endpoint_ok,
            "bpd_ok": bpd_ok,
            "structural_ok": structural_ok,
            "timeout_seconds": args.timeout,
        },
        "results": [dataclasses.asdict(r) for r in results],
    }
    json_path.parent.mkdir(parents=True, exist_ok=True)
    json_path.write_text(json.dumps(payload, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"JSON report written: {json_path}")


if __name__ == "__main__":
    main()

File: scripts/test-bank-matrix.ts
#!/usr/bin/env bun
/**
 * Top-bank FinTS anonymous-init smoke matrix.
 *
 * What it tests (no credentials required):
 * - Endpoint reachability + FinTS envelope acceptance
 * - BPD presence (HIBPA)
 * - HIKAZS / HICAZS support advertisement
 * - HITANS TAN method IDs
 *
 * Usage:
 *     bun scripts/test-bank-matrix.ts
 *     bun scripts/test-bank-matrix.ts --timeout 20
 */

import { readFileSync, writeFileSync, mkdirSync } from "fs";
import { resolve, dirname } from "path";

const PRODUCT_ID = "F7C4049477F6136957A46EC28";
const COUNTRY = "280";

interface BankInfo {
	blz: string;
	bic: string;
	name: string;
	url: string;
}

interface ProbeResult {
	label: string;
	blz: string;
	name: string;
	url: string;
	endpoint_ok: boolean;
	bpd_ok: boolean;
	structural_ok: boolean;
	hikaz_versions: string[];
	has_hicazs: boolean;
	tan_methods: string[];
	response_codes: string[];
	http_status: number | null;
	error: string;
	variant: string;
}

function buildAnonymousInit(blz: string): string {
	const hkidn = `HKIDN:2:2+${COUNTRY}:${blz}+0+0+0'`;
	const hkvvb = `HKVVB:3:3+0+0+0+${PRODUCT_ID}+1.0'`;
	const inner = hkidn + hkvvb;
	const hnhbs = "HNHBS:4:1+1'";

	const headerPrefix = "HNHBK:1:3+";
	const headerSuffix = "+300+0+1'";
	const headerLen = headerPrefix.length + 12 + headerSuffix.length;
	const total = headerLen + inner.length + hnhbs.length;
	const header = `${headerPrefix}${String(total).padStart(12, "0")}${headerSuffix}`;
	return header + inner + hnhbs;
}

function buildAnonymousInitWithHksyn(blz: string): string {
	const hkidn = `HKIDN:2:2+${COUNTRY}:${blz}+0+0+0'`;
	const hkvvb = `HKVVB:3:3+0+0+0+${PRODUCT_ID}+5.0.0'`;
	const hksyn = "HKSYN:4:3+0'";
	const inner = hkidn + hkvvb + hksyn;
	const hnhbs = "HNHBS:5:1+1'";

	const headerPrefix = "HNHBK:1:3+";
	const headerSuffix = "+300+0+1'";
	const headerLen = headerPrefix.length + 12 + headerSuffix.length;
	const total = headerLen + inner.length + hnhbs.length;
	const header = `${headerPrefix}${String(total).padStart(12, "0")}${headerSuffix}`;
	return header + inner + hnhbs;
}

function decodeFintsResponse(raw: ArrayBuffer): string {
	const text = new TextDecoder("latin1").decode(raw);
	const clean = text.replace(/[\n\r ]/g, "");
	const decoded = Buffer.from(clean, "base64");
	return new TextDecoder("latin1").decode(decoded);
}

function parseSegmentVersion(header: string): string | null {
	const parts = header.split(":");
	if (parts.length < 3) return null;
	const ver = parts[2];
	return /^\d+$/.test(ver) ? ver : null;
}

function analyzeResponse(decoded: string) {
	const segments = decoded.split("'").filter(Boolean);

	const bpd_ok = segments.some((s) => s.startsWith("HIBPA"));
	const codeMatches = decoded.match(/([0-9]{4}):/g) || [];
	const codes = [...new Set(codeMatches.map((c) => c.slice(0, 4)))].sort();
	const structural_ok = !codes.includes("9110");

	const hikaz_versions: Set<string> = new Set();
	let has_hicazs = false;
	const tan_methods: Set<string> = new Set();

	for (const seg of segments) {
		if (seg.startsWith("HIKAZS:")) {
			const header = seg.split("+", 1)[0];
			const ver = parseSegmentVersion(header);
			if (ver) hikaz_versions.add(ver);
		} else if (seg.startsWith("HICAZS:")) {
			has_hicazs = true;
		} else if (seg.startsWith("HITANS:")) {
			const body = seg.split("+").slice(1);
			for (const field of body) {
				const m = field.match(/^([0-9]{3}):[0-9A-Z]:/);
				if (m) tan_methods.add(m[1]);
			}
		}
	}

	return {
		bpd_ok,
		structural_ok,
		hikaz_versions: [...hikaz_versions].sort((a, b) => Number(a) - Number(b)),
		has_hicazs,
		tan_methods: [...tan_methods].sort(),
		codes,
	};
}

async function sendAnonymousProbe(
	bank: BankInfo,
	timeoutMs: number,
	variant: string = "anon",
): Promise<ProbeResult> {
	const base: Omit<ProbeResult, "endpoint_ok" | "bpd_ok" | "structural_ok" | "hikaz_versions" | "has_hicazs" | "tan_methods" | "response_codes" | "http_status" | "error"> = {
		label: bank.name,
		blz: bank.blz,
		name: bank.name,
		url: bank.url,
		variant,
	};

	try {
		const msg = variant === "anon_hksyn"
			? buildAnonymousInitWithHksyn(bank.blz)
			: buildAnonymousInit(bank.blz);
		const body = Buffer.from(msg, "latin1").toString("base64");

		const controller = new AbortController();
		const timer = setTimeout(() => controller.abort(), timeoutMs);

		const resp = await fetch(bank.url, {
			method: "POST",
			body,
			headers: { "Content-Type": "text/plain" },
			signal: controller.signal,
		});
		clearTimeout(timer);

		if (resp.status !== 200) {
			return {
				...base,
				endpoint_ok: false,
				bpd_ok: false,
				structural_ok: false,
				hikaz_versions: [],
				has_hicazs: false,
				tan_methods: [],
				response_codes: [],
				http_status: resp.status,
				error: `HTTP ${resp.status}`,
			};
		}

		const rawBuf = await resp.arrayBuffer();
		const decoded = decodeFintsResponse(rawBuf);
		const analysis = analyzeResponse(decoded);

		return {
			...base,
			endpoint_ok: true,
			...analysis,
			response_codes: analysis.codes,
			http_status: resp.status,
			error: "",
		};
	} catch (err) {
		return {
			...base,
			endpoint_ok: false,
			bpd_ok: false,
			structural_ok: false,
			hikaz_versions: [],
			has_hicazs: false,
			tan_methods: [],
			response_codes: [],
			http_status: null,
			error: String(err),
		};
	}
}

function chooseBetterResult(primary: ProbeResult, secondary: ProbeResult): ProbeResult {
	if (!primary.endpoint_ok && secondary.endpoint_ok) return secondary;
	if (primary.endpoint_ok && !secondary.endpoint_ok) return primary;

	const score = (r: ProbeResult) =>
		(r.bpd_ok ? 4 : 0) +
		(r.structural_ok ? 2 : 0) +
		(r.hikaz_versions.length > 0 ? 1 : 0) +
		(r.has_hicazs ? 1 : 0);

	return score(secondary) > score(primary) ? secondary : primary;
}

function parseBanksZig(path: string): Map<string, BankInfo> {
	const text = readFileSync(path, "utf-8");
	const pattern = /makeBank\("(\d{8})",\s*"([^"]*)",\s*"([^"]*)",\s*"([^"]*)"\)/g;
	const banks = new Map<string, BankInfo>();
	let m: RegExpExecArray | null;
	while ((m = pattern.exec(text)) !== null) {
		banks.set(m[1], { blz: m[1], bic: m[2], name: m[3], url: m[4] });
	}
	return banks;
}

function chooseAtruviaRepresentative(banks: Map<string, BankInfo>): BankInfo | null {
	const candidates = [...banks.values()]
		.filter(
			(b) =>
				b.url.includes("fints1.atruvia.de/cgi-bin/hbciservlet") &&
				(/Volksbank|Raiffeisenbank|VR Bank/.test(b.name)),
		)
		.sort((a, b) => a.blz.localeCompare(b.blz));
	return candidates[0] ?? null;
}

function buildTargetList(banks: Map<string, BankInfo>): Array<[string, BankInfo]> {
	const targets: Array<[string, string]> = [
		["Comdirect", "20041177"],
		["Berliner Sparkasse", "10050000"],
		["Deutsche Bank", "10070000"],
		["Commerzbank", "10040000"],
		["Postbank", "10010010"],
		["ING", "50010517"],
		["DKB", "12030000"],
	];

	const out: Array<[string, BankInfo]> = [];
	for (const [label, blz] of targets) {
		const bank = banks.get(blz);
		if (bank) out.push([label, bank]);
	}

	const atruvia = chooseAtruviaRepresentative(banks);
	if (atruvia) out.push(["Atruvia representative", atruvia]);
	return out;
}

function yn(v: boolean): string {
	return v ? "YES" : "NO";
}

function printMatrix(results: ProbeResult[]) {
	const cols: Array<[string, number]> = [
		["Bank", 24],
		["BLZ", 8],
		["Endpoint", 8],
		["BPD", 4],
		["Struct", 6],
		["HIKAZS", 8],
		["HICAZS", 6],
		["TAN methods", 18],
		["Codes", 20],
		["Variant", 10],
	];
	console.log(cols.map(([n, w]) => n.padEnd(w)).join("  "));
	console.log(cols.map(([, w]) => "-".repeat(w)).join("  "));

	for (const r of results) {
		const hikaz = r.hikaz_versions.join(",") || "-";
		const tan = r.tan_methods.join(",") || "-";
		const codes = r.response_codes.join(",") || "-";
		const row = [
			r.label.slice(0, 24).padEnd(24),
			r.blz.padEnd(8),
			yn(r.endpoint_ok).padEnd(8),
			yn(r.bpd_ok).padEnd(4),
			yn(r.structural_ok).padEnd(6),
			hikaz.slice(0, 8).padEnd(8),
			yn(r.has_hicazs).padEnd(6),
			tan.slice(0, 18).padEnd(18),
			codes.slice(0, 20).padEnd(20),
			r.variant.slice(0, 10).padEnd(10),
		];
		console.log(row.join("  "));
		if (r.error) console.log(`    error: ${r.error}`);
		console.log(`    url: ${r.url}`);
	}
}

// --- Main ---

const args = process.argv.slice(2);
let timeoutS = 15;
let tryHksyn = false;
let jsonOut = "scripts/bank-matrix-last.json";

for (let i = 0; i < args.length; i++) {
	if (args[i] === "--timeout" && args[i + 1]) {
		timeoutS = Number(args[++i]);
	} else if (args[i] === "--try-hksyn-variant") {
		tryHksyn = true;
	} else if (args[i] === "--json-out" && args[i + 1]) {
		jsonOut = args[++i];
	}
}

const repoRoot = resolve(dirname(new URL(import.meta.url).pathname), "..");
const banksPath = resolve(repoRoot, "libwimg/src/banks.zig");
const banksByBlz = parseBanksZig(banksPath);
const targets = buildTargetList(banksByBlz);

console.log(`Running anonymous FinTS probes for ${targets.length} banks (timeout=${timeoutS.toFixed(1)}s)\n`);

const results: ProbeResult[] = [];
for (const [label, bank] of targets) {
	process.stdout.write(`- probing ${label} (${bank.blz}) ...\n`);
	const primary = await sendAnonymousProbe(
		{ ...bank, name: label },
		timeoutS * 1000,
		"anon",
	);
	let chosen = primary;
	if (tryHksyn && (!primary.bpd_ok || !primary.structural_ok)) {
		process.stdout.write("  -> trying anon_hksyn variant ...\n");
		const secondary = await sendAnonymousProbe(
			{ ...bank, name: label },
			timeoutS * 1000,
			"anon_hksyn",
		);
		chosen = chooseBetterResult(primary, secondary);
	}
	results.push(chosen);
}

console.log();
printMatrix(results);

const endpointOk = results.filter((r) => r.endpoint_ok).length;
const bpdOk = results.filter((r) => r.bpd_ok).length;
const structuralOk = results.filter((r) => r.structural_ok).length;
console.log(
	`\nSummary: endpoint_ok=${endpointOk}/${results.length}, bpd_ok=${bpdOk}/${results.length}, structural_ok=${structuralOk}/${results.length}`,
);

const payload = {
	summary: {
		count: results.length,
		endpoint_ok: endpointOk,
		bpd_ok: bpdOk,
		structural_ok: structuralOk,
		timeout_seconds: timeoutS,
	},
	results,
};

const outPath = resolve(repoRoot, jsonOut);
mkdirSync(dirname(outPath), { recursive: true });
writeFileSync(outPath, JSON.stringify(payload, null, 2));
console.log(`JSON report written: ${outPath}`);
